sdk/core/azure-core-amqp/vendor/azure-uamqp-c/src/connection.c

```diff
@@ -173,49 +173,64 @@ static int channel_table_allocate(CHANNEL_TABLE* channel_table, uint16_t* h)
-              channel_table->channels[i].is_endpoint_live = false;
+                LogError("Cannot allocate more channels");
+                result = MU_FAILURE;
+            }
+            else
+            {
+                uint32_t new_capacity = channel_table->capacity + INITIAL_CHANNEL_TABLE_CAPACITY;
+                if (channel_table->capacity >= UINT16_MAX)
+                {
+                    new_capacity = UINT16_MAX;
+                }
+                CHANNEL_TABLE_ENTRY* new_channels = (CHANNEL_TABLE_ENTRY*)realloc(
+                    channel_table->channels, sizeof(CHANNEL_TABLE_ENTRY) * new_capacity);
+                if (new_channels == NULL)
+                {
+                    LogError("Cannot reallocate memory for handle table");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    channel_table->channels = new_channels;
+                    channel_table->capacity = new_capacity;
+                    // Ensure the newly resized part of the handle table is empty.
+                    uint32_t i = channel_table->size;
+                    do
+                    {
+                        channel_table->channels[i].outgoing_channel = ENDPOINT_UNUSED_CHANNEL;
+                        channel_table->channels[i].incoming_channel = ENDPOINT_UNUSED_CHANNEL;
+                        channel_table->channels[i].is_endpoint_live = false;
+                        i++;
+                    } while (i < channel_table->capacity);
+                }
             }
-            channel_table->capacity *= 2;
-          }
         }
         if (!result)
         {
-          uint16_t i = 0;
-          for (; i < channel_table->size; i++)
-          {
-            if (channel_table->channels[i].outgoing_channel == ENDPOINT_UNUSED_CHANNEL)
+            // Look for an empty slot in the table. If none is found, add to the end of the table.
+            uint32_t i = 0;
+            for (; i < channel_table->size; i++)
+            {
+                if (channel_table->channels[i].outgoing_channel == ENDPOINT_UNUSED_CHANNEL)
+                {
+                    break;
+                }
+            }
+            // If we didn't find a hole, we need to increase the size of the table by 1.
+            if (i == channel_table->size)
             {
-              *h = i;
-              channel_table->channels[i].outgoing_channel = i;
-              break;
+                channel_table->size++;
             }
-          }
-          // If we didn't find a hole, add this to the end of the table.
-          if (i == channel_table->size)
-          {
-            *h = channel_table->size;
-            channel_table->channels[channel_table->size].outgoing_channel = channel_table->size;
-            channel_table->channels[channel_table->size].is_endpoint_live = true;
-            channel_table->size++;
-          }
+
+            *h = (uint16_t)i;
```

Comment:
Shouldn't it be `*h = channel_table->size`?
I could be less by 1, if we went into condition above.

Reply:
If we aren't increasing the size of the table (if we re-use an existing slot in the table), then `channel_table->size` is the wrong value. I is either the index into the table where we found a hole, *or* the last entry in the table.

Reply:
In other words, if we take the `break` at line 221, then `i` will be an index into the middle of the table.

Reply:
Right! Thanks.